samples/nrf9160/gnss/src/assistance.c

```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2021 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+#include <zephyr.h>
+#include <modem/modem_info.h>
+#include <modem/modem_jwt.h>
+#include <net/nrf_cloud_rest.h>
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+#include <net/nrf_cloud_agps.h>
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+#include <pm_config.h>
+#include <date_time.h>
+#include <net/nrf_cloud_pgps.h>
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+#include "assistance.h"
+
+static char jwt_buf[600];
+static char rx_buf[2048];
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static char agps_data_buf[3500];
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static struct nrf_modem_gnss_agps_data_frame agps_need;
+static struct gps_pgps_request pgps_request;
+static struct nrf_cloud_pgps_prediction *prediction;
+static struct k_work get_pgps_data_work;
+static struct k_work inject_pgps_data_work;
+static K_SEM_DEFINE(time_sem, 0, 1);
+#endif /* CONFIG_NRF_CLOUD_PGPS */
+
+static struct k_work_q *work_q;
+static volatile bool assistance_active;
+
+#if defined(CONFIG_NRF_CLOUD_AGPS)
+static int serving_cell_info_get(struct lte_lc_cell *serving_cell)
+{
+	int err;
+
+	err = modem_info_init();
+	if (err) {
+		return err;
+	}
+
+	char resp_buf[MODEM_INFO_MAX_RESPONSE_SIZE];
+
+	err = modem_info_string_get(MODEM_INFO_CELLID,
+				    resp_buf,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err < 0) {
+		return err;
+	}
+
+	serving_cell->id = strtol(resp_buf, NULL, 16);
+
+	err = modem_info_string_get(MODEM_INFO_AREA_CODE,
+				    resp_buf,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err < 0) {
+		return err;
+	}
+
+	serving_cell->tac = strtol(resp_buf, NULL, 16);
+
+	/* Request for MODEM_INFO_MNC returns both MNC and MCC in the same string. */
+	err = modem_info_string_get(MODEM_INFO_OPERATOR,
+				    resp_buf,
+				    MODEM_INFO_MAX_RESPONSE_SIZE);
+	if (err < 0) {
+		return err;
+	}
+
+	serving_cell->mnc = strtol(&resp_buf[3], NULL, 10);
+	/* Null-terminate MCC, read and store it. */
+	resp_buf[3] = '\0';
+	serving_cell->mcc = strtol(resp_buf, NULL, 10);
+
+	return 0;
+}
+#endif /* CONFIG_NRF_CLOUD_AGPS */
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void get_pgps_data_work_fn(struct k_work *work)
+{
+	ARG_UNUSED(work);
+
+	int err;
+
+	printk("Sending request for P-GPS predictions to nRF Cloud...\n");
```

Comment:
The logger could be used here, though that may require the whole sample to be converted

Reply:
I didn't want to switch to logger, because I think the output is clearer without (if you consider the GNSS data which is printed once a second). The output related to A-GPS and P-GPS could use logger. Do you think I should change that?

Reply:
Hm, it could be beneficial in the case where NMEA printing is enabled, perhaps. Then the assistance printing can be turned off with the logger to avoid cluttering the output

Reply:
That's a good point. Output related to assistance is tricky, because it interferes with the GNSS PVT/NMEA output. The current way of handling this is that PVT/NMEA output is paused while assistance data is being downloaded. Logger would make it possible to suppress the assistance related output when needed, but this would need to be somehow taken into account in the PVT/NMEA output so that it isn't unnecessarily paused when assistance output is disabled. I think this could be done by checking the logging level. I'll look into this.

Reply:
My version of this sample did switch to logger, btw...  Though I didn't test using an external NMEA visualizer.  What program do you use for that @tokangas ?